src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1454,3 +1455,107 @@ int32_t SystemNative_ReadProcessStatusInfo(pid_t pid, ProcessStatus* processStat
     return -1;
 #endif // __sun
 }
+
+int32_t SystemNative_PRead(intptr_t fd, void* buffer, int32_t bufferSize, int64_t fileOffset)
+{
+    assert(buffer != NULL);
+    assert(bufferSize >= 0);
+
+    ssize_t count;
+    while ((count = pread(ToFileDescriptor(fd), buffer, (uint32_t)bufferSize, (off_t)fileOffset)) < 0 && errno == EINTR);
+
+    assert(count >= -1 && count <= bufferSize);
+    return (int32_t)count;
+}
+
+int32_t SystemNative_PWrite(intptr_t fd, void* buffer, int32_t bufferSize, int64_t fileOffset)
+{
+    assert(buffer != NULL);
+    assert(bufferSize >= 0);
+
+    ssize_t count;
+    while ((count = pwrite(ToFileDescriptor(fd), buffer, (uint32_t)bufferSize, (off_t)fileOffset)) < 0 && errno == EINTR);
+
+    assert(count >= -1 && count <= bufferSize);
+    return (int32_t)count;
+}
+
+int64_t SystemNative_PReadV(intptr_t fd, IOVector* vectors, int32_t vectorCount, int64_t fileOffset)
+{
+    assert(vectors != NULL);
+    assert(vectorCount >= 0);
+
+    int64_t count = 0;
+    int fileDescriptor = ToFileDescriptor(fd);
+#if HAVE_PREADV && !defined(TARGET_WASM) // preadv is buggy on WASM
```

Comment:
Is there an issue to go along with that bug?

Reply:
@stephentoub I've not reported it yet.

@vargaz when I was testing `preadv` and `pwritev` for WASM, our unit tests were failing (while there were passing on the same Ubuntu machine for coreclr built). It seemed that when providing n-buffers and offset=0 to preadv|pwritev, all  read|write operations were happening at offset 0 (while only the first should, every next operation should start where the previous has finished). I've decided to simply not to use them on WASM (hence `!defined(TARGET_WASM) `) and instead fallback to our own implementation that uses `pread` and `pwrite` (which we need for macOS anyway, as only macOS 11+ provides `preadv` and `pwritev`)

The most suprising thing is that it happens in the native layer ([source code](https://github.com/dotnet/runtime/blob/05327b18be5b832089866d2326908192ae433c60/src/libraries/Native/Unix/System.Native/pal_io.c#L1483-L1521))

@vargaz where should I report the bug?